[Support] [lldb] Fix thread jump #45326

da-viper opened the pull request:
Fixes #45326 

When you thread jump by calling 
`j +2` or `thread jump --by +2`  the offset is not recognised. This commit fixes that.

da-viper committed:
[lldb] Add test for jumping by offset

Signed-off-by: Ebuka Ezike <yerimyah1@gmail.com>

da-viper committed:
[lldb] Add test main.cpp file

Signed-off-by: Ebuka Ezike <yerimyah1@gmail.com>

dwblaikie commented:
Looks like this could/should be 2-3 comits. The ADT change is one, the llvm-symbolizer could be another (with or without the lldb test coverage), then possibly the lldb test coverage as a separate third step. (the contents of the patch I don't have much opinion on, I haven't even thought about whether the ADT change is good/bad, etc)

da-viper commented:
I wasn't sure if I should split it as they are quite small on their own. 

Will create a two new PR for the `symbolizer` and `llvm::StringRef` and this PR will depend on the llvm::StringRef one.

da-viper committed:
[lldb][lldb-dap] remove `+` prefix

da-viper committed:
[lldb][lldb-dap] remove `+` prefix

checking if it starts_with prefix is redundant since consume front also check it

ashgti commented on lldb/source/Commands/CommandObjectThread.cpp:
```diff
@@ -1649,11 +1649,14 @@ class CommandObjectThreadJump : public CommandObjectParsed {
         break;
-      case 'b':
+      case 'b': {
+        option_arg.consume_front("+");
```
Should this be documented that an optional `+` can be specified?

ashgti replied:
I was just think that https://github.com/llvm/llvm-project/blob/88c4ef2f9fc0cda90c8452bc1c46844aaa722a3e/lldb/source/Commands/Options.td#L1139 doesn't mention that the leading `+` is supported or that the value can be negative. Also I can't tell, at least from the help text, if this includes whitespace or not. 

Like if I had:
```
  int a = 3;
  a = 1;

  a = 2; //<pc=here> thread jump -b -1 does this go to the blank line?
```

da-viper replied:
added documentation clarifying it can be negative or positive.

ashgti commented on lldb/test/API/functionalities/thread/jump/TestThreadJump.py:
```diff
@@ -62,6 +65,71 @@ def test(self):
+            ],
+        )
+
+        self.runCmd(f"thread jump --by +{pos_jump_offset}")
```
If the `+` prefix is optional, should we also test that this works without the prefix?

da-viper replied:
it is tested in the previous test_function

ashgti reviewed (commented):
I'm less familiar with this part of lldb, so I'd defer to others here.

da-viper committed:
[lldb] update thread jump documentation.
